torch/_refs/__init__.py

```diff
@@ -2961,6 +2962,66 @@ def diagonal(
     return result
 
 
+@register_decomposition(torch.ops.aten.diag_embed)
+def diag_embed(
+    self: TensorLikeType,
+    offset: int = 0,
+    dim1: int = -2,
+    dim2: int = -1,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.diag_embed
+    """
+    # as per the docs, exchanging dims is equivalent to changing the sign of
+    # offset
+    if dim1 > dim2:
+        dim1, dim2 = dim2, dim1
+        offset = -offset
+
+    # convert from negative dims
+    rank = self.ndim + 1
+    dim1 = utils.canonicalize_dim(rank=rank, idx=dim1)
+    dim2 = utils.canonicalize_dim(rank=rank, idx=dim2)
+
+    check(
+        dim1 != dim2, lambda: f"diagonal dimensions cannot be identical {dim1}, {dim2}"
+    )
+
+    # insert padding if offset is not 0, increase the size of last dim, then
+    # roll to align the data to match 1s in eye
+    if offset != 0:
+        z_shape = list(self.shape)
+        z_shape[-1] = builtins.abs(offset)
+        z = torch.zeros(
+            z_shape, dtype=self.dtype, device=self.device, requires_grad=False
+        )
+        self = torch.cat((self, z), dim=-1)
+        self = torch.roll(self, self.size(-1) if offset < 0 else offset, dims=-1)
+
+    # as per the docs, the size of last dim is placed at dim1 and dim2
+    last_dim = self.size(-1)
+
+    # preserve original data, but place 1 at dim1 and move last dim to dim2
+    # TODO: use movedim here once it's available:
+    # a = self.unsqueeze(dim1).movedim(-1, dim2)
```

Comment:
What about implementing it as part of this stack. It seems like it's just a composite of `permute`, isn't it?